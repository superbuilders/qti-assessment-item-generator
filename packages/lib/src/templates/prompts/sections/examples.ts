import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
