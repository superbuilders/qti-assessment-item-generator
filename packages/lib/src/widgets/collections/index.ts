import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
