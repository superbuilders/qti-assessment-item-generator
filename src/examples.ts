import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
