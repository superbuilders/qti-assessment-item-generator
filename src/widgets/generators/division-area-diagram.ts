import { z } from "zod";
import { theme } from "../../utils/theme";
import type { WidgetGenerator } from "../types";

// Using factories to prevent Zod from creating schemas with $refs, which can cause issues with some tools.

/**
 * Defines a value for the quotient row, which can either be a known string value
 * or a placeholder box indicating a missing number to be filled in.
 */
const createValueOrPlaceholderSchema = () => z.discriminatedUnion("type", [
    z.object({
        type: z.literal("value").describe("A known numeric value."),
        value: z.number().describe("The numeric value to display (e.g., 10, 7).")
    }),
    z.object({
        type: z.literal("placeholder").describe("An unknown value represented by an empty box.")
    })
]).describe("Represents either a known value or a placeholder box.");

/**
 * Defines a single calculation step within a column of the area model, including
 * the numbers to be subtracted and the resulting difference.
 */
const createStepSchema = () => z.object({
    subtrahend: z.number().nullable().describe("The value being subtracted in this step (e.g., 30). Null to hide."),
    difference: z.number().nullable().describe("The result of the subtraction (e.g., 21). Null to hide."),
    differenceColor: z.string().nullable().describe("Optional CSS color for the difference text (e.g., 'red'). Null for default color.")
}).strict();

/**
 * Main Zod schema for the divisionAreaDiagram widget.
 *
 * This widget creates a visual area model for long division, commonly used in elementary
 * and middle school math to teach the partial quotients method. The diagram is structured
 * like a box, with the divisor on the left, the dividend inside, and partial quotients
 * along the top. Each step of the division process is shown as a subtraction within a
 * column of the model.
 *
 * The schema is designed to be flexible, allowing for the rendering of problems at
 * various stages of completion, including initial setup with placeholders, intermediate
 * steps with calculations, and the final solved state.
 */
export const DivisionAreaDiagramPropsSchema = z.object({
    type: z.literal("divisionAreaDiagram"),
    divisor: z.number().describe("The divisor, displayed on the left side of the model."),
    dividend: z.number().describe("The initial dividend, displayed in the first column."),
    quotientParts: z.array(createValueOrPlaceholderSchema()).describe("An array of quotient parts displayed across the top."),
    steps: z.array(createStepSchema()).describe("An array of calculation steps. Its length must match quotientParts."),
    showFinalRemainderBox: z.boolean().default(false).describe("If true, shows the final zero remainder with a box around it.")
}).strict().describe("Creates a visual area model for long division, showing partial quotients and step-by-step subtraction.");

export type DivisionAreaDiagramProps = z.infer<typeof DivisionAreaDiagramPropsSchema>;

/**
 * Generates an HTML table representing a division area model.
 */
export const generateDivisionAreaDiagram: WidgetGenerator<typeof DivisionAreaDiagramPropsSchema> = async (props) => {
    const { divisor, dividend, quotientParts, steps, showFinalRemainderBox } = props;

    // Helper to render string content, optionally with a color.
    const renderContent = (content: string | number | null, color?: string | null): string => {
        if (content === null) return '&nbsp;';
        const text = typeof content === "number" ? String(content) : content;
        // Simple HTML escaping for numbers/operators
        const escapedContent = text.replace(/</g, "&lt;").replace(/>/g, "&gt;");
        const style = color ? `color: ${color};` : '';
        // Using a span to ensure styling is applied correctly.
        return `<span style="${style}">${escapedContent}</span>`;
    };

    // Helper to render a placeholder box.
    const renderPlaceholder = (): string => {
        return `<div style="width: 40px; height: 24px; border: 1.5px solid ${theme.colors.border}; margin: auto;"></div>`;
    };
    
    // Main table wrapper
    let html = `<div style="display: inline-block; font-family: ${theme.font.family.sans}; font-size: 1.2em;">`;
    html += `<table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;">`;
    html += '<tbody>';

    // Top Row: Quotient Parts
    html += '<tr>';
    // Top-left corner spacer (no borders to avoid visible corner artifact)
    html += `<td style="width: 40px; height: 40px; border: none; padding: 0;"></td>`;
    quotientParts.forEach(part => {
        const content = part.type === 'value' ? renderContent(part.value) : renderPlaceholder();
        const style = `width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;`;
        html += `<td style="${style}">${content}</td>`;
    });
    html += '</tr>';

    // Main Row: Divisor and Steps
    html += '<tr>';
    
    // Divisor Cell
    html += `<td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;">`;
    html += `<div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;">${renderContent(divisor)}</div>`;
    html += `</td>`;

    // Steps Cells
    let currentDividend = dividend;
    steps.forEach((step, index) => {
        const isFirst = index === 0;
        const isLast = index === steps.length - 1;
        const borderStyle = `border-top: 1.5px solid black; border-bottom: 1.5px solid black;${isFirst ? " border-left: 1.5px solid black;" : ""}${isLast ? " border-right: 1.5px solid black;" : ""}`;
        html += `<td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; ${borderStyle}">`;
        // Inner table for this step's calculation
        html += `<table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody>`;
        
        // Dividend for this step
        html += `<tr><td colspan="2" style="text-align:center; padding: 8px 5px;">${renderContent(currentDividend)}</td></tr>`;
        
        if (step.subtrahend !== null) {
            // Subtraction line
            html += `<tr style="border-bottom: 1.5px solid black;">`;
            html += `<td style="text-align:right; padding: 2px 2px 2px 0;">${renderContent('âˆ’')}</td>`;
            html += `<td style="text-align:left; padding: 2px 0 2px 2px;">${renderContent(step.subtrahend)}</td>`;
            html += `</tr>`;

            if (step.difference !== null) {
                // Difference (remainder)
                html += `<tr><td colspan="2" style="text-align:center; padding: 8px 5px;">`;
                if (showFinalRemainderBox && index === steps.length - 1 && step.difference === 0) {
                     html += `<div style="border: 1.5px solid black; display: inline-block; padding: 1px 8px;">${renderContent(step.difference)}</div>`;
                } else {
                    html += renderContent(step.difference, step.differenceColor);
                }
                 html += `</td></tr>`;
            }
        }

        html += `</tbody></table></td>`;
        
        // The difference of this step becomes the dividend of the next.
        if (step.difference !== null) {
            currentDividend = step.difference;
        }
    });

    html += '</tr>';
    html += '</tbody></table></div>';
    
    return html;
};