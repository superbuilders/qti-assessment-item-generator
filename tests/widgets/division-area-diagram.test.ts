import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import { divisionAreaDiagramExamples_51_div_3, divisionAreaDiagramExamples_28_div_2 } from "../../examples/division-area-diagram"
import type { Widget } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: division-area-diagram", () => {
  const examples = [
    ...divisionAreaDiagramExamples_51_div_3,
    ...divisionAreaDiagramExamples_28_div_2
  ] as unknown as Widget[]

  examples.forEach((props, index) => {
    test(`should produce consistent output for example #${index + 1}`, async () => {
      const result = await errors.try(generateWidget(props))
      if (result.error) throw result.error
      expect(result.data).toMatchSnapshot()
    })
  })
})


