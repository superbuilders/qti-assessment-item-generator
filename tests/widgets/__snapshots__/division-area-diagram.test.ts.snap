// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: division-area-diagram should produce consistent output for example #1 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">3</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">51</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">51</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #2 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">3</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">51</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">30</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">21</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">21</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #3 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">3</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">51</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">30</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">21</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">21</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #4 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">7</span></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">3</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">51</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">30</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">21</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">21</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">21</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><div style="border: 1.5px solid black; display: inline-block; padding: 1px 8px;"><span style="">0</span></div></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #5 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">2</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">28</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">28</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #6 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">2</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">28</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">20</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">8</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">8</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #7 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><div style="width: 40px; height: 24px; border: 1.5px solid #e0e0e0; margin: auto;"></div></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">2</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">28</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">20</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">8</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">8</span></td></tr></tbody></table></td></tr></tbody></table></div>"`;

exports[`Widget: division-area-diagram should produce consistent output for example #8 1`] = `"<div style="display: inline-block; font-family: sans-serif; font-size: 1.2em;"><table cellpadding="10" align="center" style="border: none; border-collapse: collapse; margin: auto;"><tbody><tr><td style="width: 40px; height: 40px; border: none; padding: 0;"></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">10</span></td><td style="width: 80px; height: 40px; background-color: #EBF5FB; text-align:center; padding: 10px;"><span style="">4</span></td></tr><tr><td style="vertical-align: top; text-align: right; padding: 0; padding-right: 5px;"><div style="padding: 12px; background-color: #EBF5FB; display: inline-block; margin-top: 5px;"><span style="">2</span></div></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-left: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">28</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">20</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="color: red;">8</span></td></tr></tbody></table></td><td style="width: 80px; vertical-align: top; text-align: center; padding: 5px; border-top: 1.5px solid black; border-bottom: 1.5px solid black; border-right: 1.5px solid black;"><table cellpadding="0" align="center" style="width: 100%; border: none; border-collapse: collapse;"><tbody><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><span style="">8</span></td></tr><tr style="border-bottom: 1.5px solid black;"><td style="text-align:right; padding: 2px 2px 2px 0;"><span style="">−</span></td><td style="text-align:left; padding: 2px 0 2px 2px;"><span style="">8</span></td></tr><tr><td colspan="2" style="text-align:center; padding: 8px 5px;"><div style="border: 1.5px solid black; display: inline-block; padding: 1px 8px;"><span style="">0</span></div></td></tr></tbody></table></td></tr></tbody></table></div>"`;
